infra/build/functions/update_build_status.py

```diff
@@ -24,71 +24,128 @@
 class MissingBuildLogError(Exception):
   """Missing build log file in cloud storage."""
 
 
 # pylint: disable=no-member
-def get_last_build(build_ids):
+def get_build_history(build_ids):
   """Returns build object for the last finished build of project."""
   credentials, image_project = google.auth.default()
   cloudbuild = build('cloudbuild',
                      'v1',
                      credentials=credentials,
                      cache_discovery=False)
 
+  history = []
+  last_build_status = None
+
   for build_id in reversed(build_ids):
     project_build = cloudbuild.projects().builds().get(projectId=image_project,
                                                        id=build_id).execute()
     if project_build['status'] == 'WORKING':
       continue
 
+    if not last_build_status:
+      last_build_status = project_build['status']
+
     if not builds_status.upload_log(build_id):
       log_name = 'log-{0}'.format(build_id)
       raise MissingBuildLogError('Missing build log file {0}'.format(log_name))
 
-    return project_build
+    success = builds_status.is_build_successful(project_build)
+
+    history.append({
+        'build_id': build_id,
+        'finish_time': project_build['finishTime'],
+        'success': success
+    })
 
-  return None
+    if len(history) == MAX_BUILD_LOGS:
+      break
 
+  return history, last_build_status
 
+
+# pylint: disable=too-many-locals
 def update_build_status(build_tag, status_filename):
   """Update build statuses."""
+  build_history = []
+  last_successful_build = []
   statuses = {}
   successes = []
   failures = []
+  not_yet_built = []
   for project_build in BuildsHistory.query(
-      BuildsHistory.build_tag == build_tag):
-    last_build = get_last_build(project_build.build_ids)
-    if not last_build:
-      logging.error('Failed to get last build for project %s',
-                    project_build.project)
+      BuildsHistory.build_tag == build_tag).order('project'):
+
+    history, last_build_status = get_build_history(project_build.build_ids)
+
+    if not last_build_status:
+      not_yet_built.append({'name': project_build.project, 'history': history})
       continue
 
-    if last_build['status'] == 'SUCCESS':
+    if last_build_status == 'SUCCESS':
       statuses[project_build.project] = True
-      successes.append({
-          'name': project_build.project,
-          'build_id': last_build['id'],
-          'finish_time': last_build['finishTime'],
-          'success': True,
-      })
+      successes.append({'name': project_build.project, 'history': history})
     else:
       statuses[project_build.project] = False
-      failures.append({
-          'name': project_build.project,
-          'build_id': last_build['id'],
-          'finish_time': last_build['finishTime'],
-          'success': False,
-      })
-
-  builds_status.upload_status(successes, failures, status_filename)
+      failures.append({'name': project_build.project, 'history': history})
```

Comment:
Doesn't seem like there's much reason to keep separate `successes` and `failures` lists anymore, as we don't separate them in the output json anymore. We can simplify all of this by no longer caring about this distinction, as it's handled in the frontend now.